File: Medium_Level_using_Cpp/Array_and_String/238. Product of Array Except Self.cpp
// Problem: 238. Product of Array Except Self
// Link: https://leetcode.com/problems/product-of-array-except-self/description/
// Time Complexity: O(n)
// Space Complexity: O(1)
// Topics: Array, Prefix Sum
// Solution:
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int totalProduct = 1;
        int countZeros = 0;
        int numsLen = nums.size();
        vector<int> answer(numsLen, 0);

        for(int i = 0; i < numsLen; i++){
            if(nums[i] != 0){
                totalProduct *= nums[i];
            } else {
                countZeros++;
            }
        }

        if(countZeros > 1){
            return answer;
        } else if(countZeros == 1){
            for(int i = 0; i < numsLen; i++){
                if(nums[i] != 0){
                    answer[i] = 0;
                } else {
                    answer[i] = totalProduct;
                }
            }
        } else {
            for(int i = 0; i < numsLen; i++){
                if(nums[i] != 0){
                    answer[i] = (totalProduct / nums[i]);
                } else {
                    answer[i] = totalProduct;
                }
            }
        }
        return answer;

    }
};

File: Medium_Level_using_Python/Arrays_and_Strings/238. Product of Array Except Self.py
# Problem: 238. Product of Array Except Self
# Link: https://leetcode.com/problems/product-of-array-except-self/description/
# Time Complexity: O(n)
# Space Complexity: O(1)
# Topics: Array, Prefix Sum
# Solution:
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        numsLen = len(nums)
        answer = [0] * numsLen
        countZeros = 0
        totalProduct = 1

        for i in range(0, numsLen):
            if nums[i] != 0:
                totalProduct *= nums[i]
            else:
                countZeros += 1

        if countZeros > 1:
            return answer
        elif countZeros == 1:
            for i in range (0, numsLen):
                if nums[i] == 0:
                    answer[i] = totalProduct
        else:
            for i in range (0, numsLen):
                answer[i] = int(totalProduct/nums[i])
        
        return answer